### ISSUE TITLE
Lint suggesting use of TryFrom for checked integer conversion

### ISSUE THREAD

**Issue by sfackler** [rust-lang/rust-clippy#3947]
Checked integer conversion code has historically looked something like this:

```rust
if foo < i32::max_value() as usize {
    thing_a
} else {
    thing_b
}
```

But now that TryFrom is stable, it can be done in a more foolproof way like this:

```rust
match i32::try_from(foo) {
    Ok(foo) => thing_a,
    Err(_) => thing_b,
}
```

It'd be cool if there was a lint guiding users to migrate their code.

**Comment by pJunger** [comment]
So now that the simple check is merged, the next steps that I can see would be something like this:

1. Further transformations
    1. Transform simple casts
    2. Transform more involved cases?

2. Detect (probably) incorrect casts like `(i < ITYPE::max_value()) && (i >= ITYPE::min_value())` ?

### Regarding 1:

Simple transformations:
```
if (i <= ITYPE::max_value()) && (i >= ITYPE::min_value()) {
    ...(i as ITYPE)... i
}
```
or a transformation for the current rustfix solution
```
if ITYPE::try_from(i).is_ok()  {
    ...(i as ITYPE)... i
}
```
into
```
if let Some(i_as_ITYPE) = ITYPE::try_from(i) {
    ...(i_as_ITYPE)... i
}
```

#### Questions:
* Is there a way to get all usages of `i` that are in scope?
* How can you emit multiple related findings (that are resolved by rustfix in one step?)

### Regarding 2:

#### Questions:
* Do we want a lint that will emit false positives?
* Or should code like this be transformed into something with more intent like:
```
match ITYPE::try_from(i) {
    Some(i_as_ITYPE) if i_as_ITYPE != ITYPE::max_value() => ...
    _ => {}
```

What do you think?
